src/compiler.ts

```diff
@@ -2001,6 +2001,54 @@ export class Compiler extends DiagnosticEmitter {
     return i64_add(memorySegment.offset, i64_new(program.runtimeHeaderSize));
   }
 
+  // Required for Closures: Allocate a function object at Runtime which is a copy of the runtime function
+  // created statically in ensureRuntimeFunction but with a possibly non-null env that holds closed
+  // over context
+  // DuncanUszkay1 TODO: Populate the env with the closed over context
+  allocFunction(instance: Function): ExpressionRef {
+    var program = this.program;
+    var module = this.module;
+    var options = this.options;
+
+    var rtInstanceOffset = this.ensureRuntimeFunction(instance);
+
+    // Create a new instance of the function object in memory
+    var rtInstance = assert(this.resolver.resolveClass(program.functionPrototype, [ instance.type ]));
+    var currentType = this.currentType;
+    var functionInstanceExpr = this.makeAllocation(rtInstance)
+    this.currentType = currentType;
```

Comment:
Because `makeAllocation` changes the current type into `Function` without any signature, we need to discard that type.

I'm not entirely sure what the desired behaviour is here. We discard the type in `ensureRuntimeFunction` by just returning an offset,  so I'm making the assumption here that we'll want to do the same.